RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,336 @@
+
+## Unit constraints of a Modelica model
+
+The following rules describe the unit meta-expression of a given Modelica expression, where it makes sense.
+They also provide the constraints that arise from different Modelica constructs.
+
+When it is specified that _`u1` must be equivalent to `u2`_, the corresponding constraint should be collected.
+
+In this section, the unit meta-expression of expression `e` is called `e.unit` and conditions of conditional constraints are highlighted with [].
+
+### Variables
+If `var` has a declared `unit`-attribute, `var.unit` must be equivalent to it.
+
+`var.unit` must be equivalent to the unit of the following attributes if present:
+ * `start`
+ * `min`
+ * `max`
+ * `nominal`
+
+(`var.unit` should be convertible to the `displayUnit`-attribute of `var` if it is present, but there is no corresponding constraints for the unit inference.)
+
+Consider the expression `e` of the form `var`, where `var` is a Modelica variable.
+* If `var` doesn't have a declared unit attribute, and has constant component variablity, `e.unit` is the `empty` unit.
+* Otherwise, `e.unit` is `var.unit`.
+
+### Literals
+`Real` literals have `empty` unit.
+
+### Binary operations
+Consider an expression `e` of the form `e1 op e2`.
+
+#### Additive operators
+* `e.unit` must be equivalent to `e1.unit`.
+* `e.unit` must be equivalent to `e2.unit`.
+
+Note that this requires introducing an intermediate variable to represent `e.unit`.
+This allows unit propagation when one of the operand has `empty` unit.
+
+#### Multiplication
+`e.unit` is `e1.unit * e2.unit`.
+
+#### Division
+`e.unit` is `e1.unit * (e2.unit ^ -1)`.
+
+#### Relational operators
+`e1.unit` must be equivalent to `e2.unit`
+
+#### Power operator
+* `e.unit` must be equivalent to `e1.unit ^ e2`.
+* If [`e2` wasn't evaluated or is a `Real` expression], `e1.unit` must be equivalent to `"1"`.
+
+Note that this requires introducing an intermediate variable to represent `e.unit`.
+Technically, `e2` is a Modelica expression rather than a rational number, which means that `e1.unit ^ e2` cannot be represented as a unit meta-expression.
+In practice, if the condition of the second constraint is verified, `e1.unit ^ e2` is `"1"` and if it isn't verified, then it can be represented as a meta-expression.
+
+(If the exponent was not evaluated or is a `Real`, the unit of the base should be `"1"`.)
+(It is up for discussion, when `e2` is `Real`, whether `e2.unit` should be equivalent to `"1"`. It could be handled in a similar way as transcendental functions. See below.)
+
+### Function calls
+Consider the expression `e` of the form `f(e1, e2, …, en)`.
+* If the output of the function has a declared unit, `e.unit` is the declared unit.
+* `e.unit` is `undefined` otherwise.
+
+If function input `i` has a declared unit, `ei.unit` must be equivalent to this unit.
+
+The natural generalization to functions with multiple outputs applies.
+
+### Transcendental functions
```

Comment:
Note that this will give different rules for the inputs to sin and Modelica.Math.sin.

Reply:
Will it?  What I see is that both variants would accept both empty unit arguments and arguments of unit equivalent to `"1"`.  The difference I see is that there is a specification for the unit of `sin(1.0'rad')`, whereas `Modelica.Math.sin(1.0'rad')` has undefined unit for now.

Once we get the chance to work on unit inference rules for functions, I hope we can make sure that `Modelica.Math.sin(1.0'rad')` gets unit `"1"`.  Then, the only remaining difference will be that `Modelica.Math.sin(angle)` can be used to infer that `angle` has unit `"rad"`, while `sin(angle)` doesn't impose a unit on `angle`.

Reply:
Why not improve it so that the rules are the same?

Reply:
Because I imagine that we want the built-in functions to be usable in unit-free expressions like this (it is not strictly necessary to think of `sin` as a function for trigonometry):
```
Real x(unit = "m") = 1.0 + sin(1.0);
```

At the same time, I think we would like this to be a unit error as long as the input of `Modelica.Math.sin` has a declared unit:
```
Real x(unit = "m") = 1.0 + Modelica.Math.sin(1.0);
```

However, I think it is too early to think deeply about how unit checking for the non-builtin functions should behave.